cluster-autoscaler/proposals/gpu_optimization.md

```diff
@@ -0,0 +1,88 @@
+# Cluster Autoscaling Optimization for GPU clusters
+##### Author: Jeffwan
+
+## Introduction
+Cluster Autoscaler makes it extremely easy to scale kubernetes cluster in response to pod status. At the same time, having it automate Kubernetes cluster to work efficiently.
+
+However, there's still challenges in specific scenarios for accelerator computing clusters. Accelerator computing instances are quite different from normal instance. They are powerful, scalable instances that provide GPU-based parallel computing capabilities. 
+
+Accelerator computings nodes are expensive. (ec2 p3.16xlarge on demand instance cost $24.48/hr). Efficiently scale the cluster will save lots of money. GPU nodes are well suited for tasks that have heavy computation needs like machine learning and HPC applications. They can not be interrupted in some cases and that brings us more challenges in scale down. 
+
+Here's the problems I find in current upstream cluster autoscaler. 
+
+### Problems in cluster autoscaler for GPU node
+
+#### Accelerator Label 
+GKE uses gpu label `cloud.google.com/gke-accelerator` to inspect if a node is an accelerator node, other cloud providers don't define their label yet. In order for CA to behavior normally in GPU cluster, it's essential to have this gpu label.
+
+#### Scale too much without GPU label support
+Pending pods will trigger scale up decision to bring up number of nodes in one node group. When new node is added, CA knows it is `upcoming` and it knows some pods will go on that node once it boots up. So it won't trigger another scale-up for those pods. Device plugin triggers after the node becomes ready and the extra resource required by pod will show up in node spec even later. CA will see that it was wrong and the pods for which it triggered scale-up can't actually go to a node it added for them (because it doesn't have the special resource yet and CA doesn't know it will be added later). So CA will go and create more nodes.
```

Comment:
> CA will see that it was wrong and the pods for which it triggered scale-up can't actually go to a node it added for them (because it doesn't have the special resource yet and CA doesn't know it will be added later). So CA will go and create more nodes.

Not really, CA will look at template node for a node group to verify if the node is expected to have GPUs. If so, it'll wait for them and not add surplus nodes.

Reply:
Yeah. That's different cases or phases.
What you mentioned are two cases 
1. nodeGroup is not GPU family but job requires GPU, scaling up is not helpful. 
2.nodeGroup is GPU family and job requires GPU, CA will check upcoming nodes and don't add surplus node until node is ready. 

The situation I meet is second one, the problem is GPU takes time to be allocatable.  When then new create node is ready, but GPU resource is not ready, autoscaler will think it's unschedulable and create new nodes.  This is hacked in GKE but not other cloud providers yet.

Reply:
> nodeGroup is not GPU family but job requires GPU, scaling up is not helpful.

This shouldn't happen if GPU resource of nodes (or node template, if a group is empty) is set correctly.

>  This is hacked in GKE but not other cloud providers yet.

To be fair, I think there were successful user reports of applying GKE label to make it work elsewhere ;) But I agree it's not perfect, we could probably move it to a constant in cloud provider or sth.

Reply:
> > nodeGroup is not GPU family but job requires GPU, scaling up is not helpful.
> 
> This shouldn't happen if GPU resource of nodes (or node template, if a group is empty) is set correctly. EDIT: do you mean that without the label, node with unready GPU can masquerade as non-GPU node?
> No. As you said, this shouldn't happen. I mean user schedules GPU workloads on non-GPU cluster. CA has support to avoid scaling up in this case because it can not fit into this kind of nodeGroup. 

>  
>
> > This is hacked in GKE but not other cloud providers yet.
> 
> To be fair, I think there were successful user reports of applying GKE label to make it work elsewhere ;) But I agree it's not perfect, we could probably move it to a constant in cloud provider or sth.

Yes. Definitely, I think all cloud providers can hack this way. But each has its own GPU label. Either support custom GPU label from arguments or move to cloud providers (preferred.)